src/Uno.UI.Runtime.Skia.Gtk/ApplicationModel/DataTransfer/ClipboardExtensions.cs

```diff
@@ -0,0 +1,336 @@
+
+					builder.AppendLine("copy");
+					foreach (var item in items)
+					{
+						var path = item.Path;
+						builder.AppendLine(UrlEncode(path));
+					}
+
+					nativeData.Set(GnomeCopiedFilesContent, 8, Encoding.UTF8.GetBytes(builder.ToString().Trim()));
+				}
+				else if (format == StandardDataFormats.ApplicationLink)
+				{
+					uris.Add((await data.GetApplicationLinkAsync()).ToString());
+				}
+				else if (format == StandardDataFormats.WebLink)
+				{
+					uris.Add((await data.GetWebLinkAsync()).ToString());
+				}
+				else if (format == StandardDataFormats.Uri)
+				{
+					uris.Add((await data.GetUriAsync()).ToString());
+				}
+
+				if (uris.Count != 0)
+				{
+					nativeData.SetUris(uris.ToArray());
+				}
+			}
+
+			uint id;
+			if (CheckFormat(StandardDataFormats.Text, out id))
+			{
+				targetList.AddTextTargets(id);
+			}
+			if (CheckFormat(StandardDataFormats.Bitmap, out id))
+			{
+				targetList.AddImageTargets(id, true);
+			}
+			if (CheckFormat(StandardDataFormats.Html, out id))
+			{
+				targetList.Add(HtmlContent, 0, id);
+			}
+			if (CheckFormat(StandardDataFormats.Rtf, out id))
+			{
+				targetList.Add(RtfContent, 0, id);
+			}
+			if (CheckFormat(StandardDataFormats.StorageItems, out id))
+			{
+				targetList.Add(GnomeCopiedFilesContent, 0, id);
+			}
+			if (CheckFormat(StandardDataFormats.ApplicationLink, out id))
+			{
+				targetList.AddUriTargets(id);
+			}
+			if (CheckFormat(StandardDataFormats.WebLink, out id))
+			{
+				targetList.AddUriTargets(id);
+			}
+			if (CheckFormat(StandardDataFormats.Uri, out id))
+			{
+				targetList.AddUriTargets(id);
+			}
+
+			_clipboard.SetWithData((TargetEntry [])targetList, SetDataNative, (clipboard) => { });
+		}
+
+		public void StartContentChanged()
+		{
+			_clipboard.OwnerChange += Clipboard_OwnerChange;
+		}
+
+		public void StopContentChanged()
+		{
+			_clipboard.OwnerChange -= Clipboard_OwnerChange;
+		}
+
+		private void Clipboard_OwnerChange(object o, OwnerChangeArgs args)
+		{
+			ContentChanged?.Invoke(this, args);
+		}
+
+		/// <summary>
+		/// Encodes a file path to a file:// Url.
+		/// While the built-in Uri class can handle this, it does not process files
+		/// such as '/home/user/%51.txt' correctly.
+		/// </summary>
+		/// <param name="path">Path to the file</param>
+		/// <returns>file:// url to the file</returns>
+		private string UrlEncode(string path)
+		{
+			var uri = new StringBuilder();
+			foreach (var ch in path)
+			{
+				if (('a' <= ch && ch <= 'z') || ('A' <= ch && ch <= 'Z') || ('0' <= ch && ch <= '9') ||
+					"-._~".Contains(ch))
+				{
+					uri.Append(ch);
+				}
+				else if (ch == Path.DirectorySeparatorChar || ch == Path.AltDirectorySeparatorChar)
+				{
+					uri.Append('/');
+				}
+				else
+				{
+					var bytes = Encoding.UTF8.GetBytes(new[] { ch });
+					foreach (var b in bytes)
+					{
+						uri.Append($"%{b:X2}");
+					}
+				}
+			}
+			return "file://" + uri;
+		}
```

Comment:
Why you're not using [`Uri.EscapeDataString`](https://docs.microsoft.com/en-us/dotnet/api/system.uri.escapedatastring)?

Reply:
https://dotnetfiddle.net/wO60Dv
It doesn't handle the DirectorySeparatorChar properly.

Reply:
We might use `Uri.EscapeDataString(str).Replace(Uri.EscapeDataString(Path.DirectorySeparatorChar), Path.DirectorySeparatorChar)`, which is a lot slower but uses the library. Should I change to that?

Reply:
@carldebilly can this be resolved?

Reply:
`new Uri("").ToString()` does what you need: https://dotnetfiddle.net/5qJNKS.

You still need to replace `\` to `/` before, though.

Reply:
@jeromelaban No, it doesn't, it's apparent in your example.

For example, for the file `/home/user/%51.txt/`, which is a perfectly valid path, to a file named `%51.txt`, the function returns `file:///home/user/Q.txt/`, which points to another file, named `Q.txt`.

The local path also returns `/home/user/Q.txt/`, which is incorrect.

This is actually the first solution I think of, but it doesn't work in this edge case.

Reply:
Got it. We'll still need to unit test this, particularly for those edge cases.

Can you move this method to a class next to Clipboard.skia.cs (e.g. FileUriHelper) and add a test here: https://github.com/unoplatform/uno/tree/master/src/Uno.UI.Tests ?

DataRows tests will be useful, similar to this one:
https://github.com/unoplatform/uno/blob/d57d8ed572184968a801dda815ac30bba96122cc/src/Uno.UI.Tests/Windows_Globalization/Given_CalendarFormatter.cs#L13-L20

Reply:
I've added the test, so I think this should be resolved for now?